api/index.js

```diff
@@ -0,0 +1,92 @@
+require('dotenv').config();
+
+const config = require('./config');
+
+const { sumVotes } = require('./utils');
+
+const { db, writeTalk, addVote, auth } = require('./db');
+
+const express = require('express');
+const bodyParser = require('body-parser')
+const app = express();
+
+const authHeaderToUserId = async authHeader => {
+  const m = authHeader && authHeader.match(/Bearer (.+)/);
+
+  if (!m) throw new Error('Request missing ID token');
+
+  const idToken = m[1];
+  const decodedToken = await auth(idToken);
+
+  return decodedToken.user_id;
+};
+
+app.use((req, res, next) => {
+  res.header('Access-Control-Allow-Origin', '*');
+  res.header('Access-Control-Allow-Headers', 'Origin, X-Requested-With, Content-Type, Accept, Authorization');
+  next();
+});
+
+app.use(bodyParser.json());
+
+app.get('/talks', async (req, res) => {
+  const limit = +req.query.limit || config.DEFAULT_PAGE_LENGTH;
+  const after = +req.query.after || 0;
+
+  const query = db.ref('talks');
+
+  const snapshot = await query.once('value');
+
+  const talks = snapshot.val();
+
+  Object.keys(talks).forEach(key => talks[key].votes = sumVotes(talks[key].votes));
+
+  res.json({
+    talks
+  });
+});
+
+app.get('/talks/:id', async (req, res) => {
+  const query = db.ref('talks/' + req.params.id);
+
+  const snapshot = await query.once('value');
+
+  const talk = snapshot.val();
+  res.json(talk);
+});
+
+app.post('/talks', async (req, res) => {
+  const { title, abstract } = req.body;
+
+  if (title.length > 50 || abstract.length > 250) return;
+
+  const userId = await authHeaderToUserId(req.header('Authorization'));
+
+  const talk = { title, abstract, userId };
+
+  writeTalk(talk);
+
+  res.json(talk);
+});
+
+app.post('/talks/:id/:vote', async (req, res) => {
+  const vals = {
+    upvote: +1,
+    downvote: -1
+  };
```

Comment:
Maybe `upvote` and `downvote` should be 2 separate fields? instead of `votes++` for an "upvote", `votes--` for an "downvote". Not a problem, I actually don't know how the real Hacker News models this.

Reply:
This way seemed a little simpler to me, though maybe it's a matter of taste.